### ISSUE TITLE
JS stack unwinding corrupts some floats?

### ISSUE THREAD

**Issue by hoodmane** [emscripten-core/emscripten#22743]
<details>
<summary>main.c++</summary>

```c++
#include <stdexcept>
#include <stdio.h>
using namespace std;

typedef unsigned int uint;
uint float_to_uint(float f);
float uint_to_float(uint i);

int main() {
    uint x = 4287976789;
    float f;
    try {
        f = uint_to_float(x);
    } catch(runtime_error& e) {
        printf("Caught runtime error...");
    }
    uint y = float_to_uint(f);
    printf("x: %ud, y: %ud, y - x: %ud\n", x, y, y - x);
    if (x != y) {
      return 1;
    }
    return 0;
}
```

</details>

<details>
<summary>helper.c++</summary>

```c++
typedef unsigned int uint;

typedef union { 
    uint i; 
    float f; 
} U; 

uint float_to_uint(float f) {
    U u;
    u.f = f;
    return u.i;
}

float uint_to_float(uint i) {
    U u;
    u.i = i;
    return u.f;
}
```

</details>

Run with:
```sh
em++ main.c++ -fexceptions -c && em++ helper.c++ -fexceptions -c &&  em++ main.o helper.o -fexceptions -o repro.js && node repro.js
```
It prints:
```console
x: 4287976789d, y: 4292171093d, y - x: 4194304d
Oops!
```
On the other hand,
```console
$ em++ main.c++ -fwasm-exceptions -c && em++ helper.c++ -fwasm-exceptions -c &&  em++ main.o helper.o -fwasm-exceptions -o repro.js && node repro.js
x: 4287976789d, y: 4287976789d, y - x: 0d
```

**Comment by kripken** [comment]
Hmm, the integer here is turned into a NaN by `uint_to_float`. And then the difference between old JS EH and new wasm EH is that the former goes through JS. Avoiding the try-catch around that call leads to the same behavior as new wasm EH, as nothing is thrown, so this is a bug in old EH (and not related to new EH).

The NaN gets canonicalized at the JS boundary, I believe (calling `invoke_fi` returns that NaN as a JS result) which is what leads to this issue. That is, we need a very particular pattern of NaN bits here, not the canonical NaN that JS (but not wasm) forces everything into.

We could in theory fix this by making `invoke_fi` return an integer and not a float directly, and we could do a reinterpret on the wasm side. That is, we could avoid sending floats through JS. I don't think that would be too much work, but OTOH this is for old EH, which is being phased out anyhow, so perhaps it isn't worth it?

**Comment by hoodmane** [comment]
Well anyone who uses rust is stuck with the old error handling until someone updates the rust try intrinsic to emit ir compatible with the new error handling:
https://github.com/emscripten-core/emscripten/issues/19612

Pyodide tries to substitute out the invokes for ones using wasm try/catch if possible so that no js frames show up and stack switching works correctly so we wouldn't be exposed to this bug. But

**Comment by kripken** [comment]
@hoodmane is that an issue on the Rust side, or LLVM? (the link is to an Emscripten issue, but I'm not sure I understood from it where the underlying problem is)

In general, getting Rust to emit IR compatible with new wasm EH sounds very important for any Rust/C++ combination, Emscripten or otherwise. Is there a Rust issue for that?

With all that said, I wouldn't be opposed to fixing float invokes for old EH, using the approach I mentioned above, if someone has time for it.

**Comment by hoodmane** [comment]
> getting Rust to emit IR compatible with new wasm EH sounds very important

Yeah I would love to switch Pyodide to use the new wasm EH and this is the only blocker.

The issue is in the rust compiler. Specifically, the `try` intrinsic needs to be updated:
https://github.com/rust-lang/rust/blob/master/compiler/rustc_codegen_llvm/src/intrinsic.rs?plain=1#L971

**Comment by kleisauke** [comment]
The tracking issue for WebAssembly EH in Rust is at https://github.com/rust-lang/rust/issues/118168. Looking at PR https://github.com/rust-lang/rust/pull/111322, I'm unsure why it was conditionally excluded for Emscripten:
https://github.com/rust-lang/rust/blob/0037048da8ed19feef7ee1437f123a88a7342b33/compiler/rustc_codegen_ssa/src/base.rs#L389-L391

Could you try to remove the `&& sess.target.os != "emscripten"` there? This would make it use `codegen_wasm_try` instead of `codegen_emcc_try`.
https://github.com/rust-lang/rust/blob/0037048da8ed19feef7ee1437f123a88a7342b33/compiler/rustc_codegen_llvm/src/intrinsic.rs#L666-L669

**Comment by kripken** [comment]
Thanks @kleisauke !

@walkingeyerobot fyi, the last comment has some details that might be relevant for future Rust/Emscripten work.

**Comment by hoodmane** [comment]
Well using that is good progress. I can get it to link with the new codegen and the try intrinsic looks perfect, but there are problems elsewhere in `rust/library/panic_unwind`. First I have to change the argument to `__cxa_begin_catch` I think it's supposed to be the wasm exception argument, the rust code for old EH dereferences it once first. This hits all the right code paths but then I'm running into the problem that `exception_header->adjustedPtr` is null. I'm not sure who was supposed to set it, I'll have to look into it more.

**Comment by hoodmane** [comment]
Looks like `adjustedPtr` was supposed to be filled in by `_Unwind_CallPersonality`, but as far as I can tell it never gets called. Was `@llvm.wasm.get.exception` supposed to call it? The catch block only has:
```wat
catch $tag0
      local.set $var4
      local.get $var3
      global.set $__stack_pointer
      local.get $var1
      local.get $var4
      local.get $var2
      call_indirect (param i32 i32)
      i32.const 1
      local.set $var5
      local.get $var5
end
```
and the `call_indirect` invokes the rust `catch_fn`.

**Comment by hoodmane** [comment]
Indeed that section is suppoed to turn into a call to `_Unwind_CallPersonality`:

> WebAssembly exception handling uses Windows exception IR for the middle level
representation. This pass does the following transformation for every
catchpad block:
(In C-style pseudocode)
> 
> - Before:
>  ```C
>   catchpad ...
>   exn = wasm.get.exception();
>   selector = wasm.get.selector();
>   ...
>  ```
> 
> - After:
> ```C
>   catchpad ...
>   exn = wasm.catch(WebAssembly::CPP_EXCEPTION);
>   // Only add below in case it's not a single catch (...)
>   wasm.landingpad.index(index);
>   __wasm_lpad_context.lpad_index = index;
>   __wasm_lpad_context.lsda = wasm.lsda();
>   _Unwind_CallPersonality(exn);
>   selector = __wasm_lpad_context.selector;
>   ...
> ```

https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/WasmEHPrepare.cpp?plain=1#L13-L33

I wonder if I can get llvm to print the result of this pass...

**Comment by hoodmane** [comment]
After the WasmEHPrepare pass the ir looks like:
```llvm
catchpad:                                         ; preds = %catchswitch
  %catchpad2 = catchpad within %catchswitch1 [ptr null]
  %exn = call ptr @llvm.wasm.catch(i32 0)
  call void %2(ptr %1, ptr %exn) [ "funclet"(token %catchpad2) ]
  catchret from %catchpad2 to label %caught
```
It seems we took that `!NeedPersonality` branch in the pass:
https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/WasmEHPrepare.cpp?plain=1#L342-L349
```c++
  // In case it is a catchpad with single catch (...) or a cleanuppad, we don't
  // need to call personality function because we don't need a selector.
  if (!NeedPersonality) {
    if (GetSelectorCI) {
      assert(GetSelectorCI->use_empty() &&
             "wasm.get.ehselector() still has uses!");
      GetSelectorCI->eraseFromParent();
    }
    return;
  }
```
So as the comment says, this would normally happen in C++ when there is a blanket `catch (...)` and so the return value of `__cxa_begin_catch` is unused. But Rust wants to know what exception was thrown. I think the simplest solution for this is for Rust to explicitly call the personality function in this case.

**Comment by hoodmane** [comment]
Adding the call to `_Unwind_CallPersonality` doesn't help because `_Unwind_GetLanguageSpecificData` returns null.

**Comment by hoodmane** [comment]
Right, the ldsa would be initialized in the ir injected by `WasmEHPrepare` if there were multiple catch blocks.

**Comment by kleisauke** [comment]
Have you already tried using [`rust/library/panic_unwind/src/gcc.rs`](https://github.com/rust-lang/rust/blob/master/library/panic_unwind/src/gcc.rs) instead of the [`rust/library/panic_unwind/src/emcc.rs`](https://github.com/rust-lang/rust/blob/master/library/panic_unwind/src/emcc.rs) implementation?
https://github.com/rust-lang/rust/blob/06d261daf62620e3449ba451898c7ad9ae750f41/library/panic_unwind/src/lib.rs#L36-L38

The former implementation is also used by `target_family = "wasm"`.

**Comment by hoodmane** [comment]
Nope, trying that now thanks.

**Comment by kleisauke** [comment]
FWIW, you may also need to adjust the following:
https://github.com/rust-lang/rust/blob/06d261daf62620e3449ba451898c7ad9ae750f41/library/unwind/src/lib.rs#L7-L10
```diff
--- a/library/unwind/src/lib.rs
+++ b/library/unwind/src/lib.rs
@@ -4,10 +4,7 @@
 #![feature(staged_api)]
 #![feature(strict_provenance)]
 #![cfg_attr(not(target_env = "msvc"), feature(libc))]
-#![cfg_attr(
-    all(target_family = "wasm", not(target_os = "emscripten")),
-    feature(simd_wasm64, wasm_exception_handling_intrinsics)
-)]
+#![cfg_attr(target_family = "wasm", feature(simd_wasm64, wasm_exception_handling_intrinsics))]
 #![allow(internal_features)]
 
 // Force libc to be included even if unused. This is required by many platforms.
```

**Comment by hoodmane** [comment]
Yup that works. Only missing thing is debug info. Emscripten has this code to attach more information to the error in debug releases:
https://github.com/emscripten-core/emscripten/blob/main/system/lib/libcxxabi/src/cxa_exception.cpp?plain=1#L299-L307
Ideally Emscripten should probably move that into `_Unwind_RaiseException()`.
